File: pages/refactorings/replace-mutable-fields-with-lenses/03-bump-oks/Before.hs
-- Bump only the successes of a batch: a hand-written map carrying
-- the branch test and the rebuild in the same step.
module Before where

data Ok     = Ok     { okValue :: Int }
  deriving (Eq, Show)
data Result = Succeeded Ok | Failed String
  deriving (Eq, Show)

bumpSucceeded :: [Result] -> [Result]
bumpSucceeded = map step
  where
    step (Succeeded ok) = Succeeded ok { okValue = okValue ok + 1 }
    step f@(Failed _)   = f

File: pages/refactorings/replace-mutable-fields-with-lenses/03-bump-oks/Before.scala
// Bump only the successes of a batch: a hand-written map carrying
// the branch test and the rebuild in the same step.
object Before:
  case class Ok(value: Int)
  enum Result:
    case Succeeded(v: Ok)
    case Failed(msg: String)

  import Result.*

  def bumpSucceeded(xs: List[Result]): List[Result] =
    xs.map {
      case Succeeded(ok) => Succeeded(ok.copy(value = ok.value + 1))
      case f @ Failed(_) => f
    }
